public/app/features/explore/state/datasource.ts

```diff
@@ -67,108 +39,35 @@ export function changeDatasource(
-      newDataSourceInstance = await getDatasourceSrv().get(datasourceName);
-    }
-
-    const currentDataSourceInstance = getState().explore[exploreId].datasourceInstance;
-    const queries = getState().explore[exploreId].queries;
     const orgId = getState().user.orgId;
+    const { history, instance } = await loadAndInitDatasource(orgId, datasourceName);
 
     dispatch(
       updateDatasourceInstanceAction({
         exploreId,
-        datasourceInstance: newDataSourceInstance,
+        datasourceInstance: instance,
+        history,
       })
     );
 
+    const queries = getState().explore[exploreId].queries;
+    const currentDataSourceInstance = getState().explore[exploreId].datasourceInstance;
+
     if (options?.importQueries) {
-      await dispatch(importQueries(exploreId, queries, currentDataSourceInstance, newDataSourceInstance));
+      await dispatch(importQueries(exploreId, queries, currentDataSourceInstance, instance));
     }
 
     if (getState().explore[exploreId].isLive) {
       dispatch(changeRefreshInterval(exploreId, RefreshPicker.offOption.value));
     }
 
-    await dispatch(loadDatasource(exploreId, newDataSourceInstance, orgId));
-
     // Exception - we only want to run queries on data source change, if the queries were imported
     if (options?.importQueries) {
       dispatch(runQueries(exploreId));
     }
   };
 }
 
-/**
- * Loads all explore data sources and sets the chosen datasource.
- * If there are no datasources a missing datasource action is dispatched.
- */
-export function loadExploreDatasourcesAndSetDatasource(
-  exploreId: ExploreId,
-  datasourceName: string
-): ThunkResult<void> {
-  return async dispatch => {
-    const exploreDatasources = getExploreDatasources();
-
-    if (exploreDatasources.length >= 1) {
-      await dispatch(changeDatasource(exploreId, datasourceName, { importQueries: true }));
-    } else {
-      dispatch(loadDatasourceMissingAction({ exploreId }));
-    }
-  };
-}
-
-/**
- * Datasource loading was successfully completed.
- */
-export const loadDatasourceReady = (
-  exploreId: ExploreId,
-  instance: DataSourceApi,
-  orgId: number
-): PayloadAction<LoadDatasourceReadyPayload> => {
-  const historyKey = `grafana.explore.history.${instance.meta?.id}`;
-  const history = store.getObject(historyKey, []);
-  // Save last-used datasource
-
-  store.set(lastUsedDatasourceKeyForOrgId(orgId), instance.name);
-
-  return loadDatasourceReadyAction({
-    exploreId,
-    history,
-  });
-};
-
-/**
- * Main action to asynchronously load a datasource. Dispatches lots of smaller actions for feedback.
- */
-export const loadDatasource = (exploreId: ExploreId, instance: DataSourceApi, orgId: number): ThunkResult<void> => {
-  return async (dispatch, getState) => {
-    const datasourceName = instance.name;
-
-    // Keep ID to track selection
-    dispatch(loadDatasourcePendingAction({ exploreId, requestedDatasourceName: datasourceName }));
-
-    if (instance.init) {
```

Comment:
As the `init` isn't async it probably does not have much value in having loading state. As the init isn't async, if it takes long time, it blocks the js loop so nothing would work, at that moment showing loading state isn't the most important thing. So init will probably be quick and the loading indicator should not be necessary.